Add information about roles in the main codebase (#135)
NUbots/NUbook: src/book/02-system/01-foundations/01-overview.mdx

All these modules are connected using NUClear which is a message passing software architecture. More can be found on NUClear [here](https://nuclear.readthedocs.io/en/latest/).

![Flow of the NUbots codebase](./images/software-overview.svg)

## Roles

> Roles in the main codebase are the building blocks for creating a binary to run on a robot. A role file lists all the modules to be used in that binary. If a module is not included in the role file, that module will not run. All roles can be found in the [roles folder](https://github.com/NUbots/NUbots/tree/master/roles) in the first level of the repository and have the extension `.role`. For specifics about the NUClear roles system, see the [NUClear page](/system/foundations/nuclear#nuclear-roles).

Reviewer: They're not building blocks, their more like a recipe.

The modules are the building blocks.

Reviewer: > If a module is not included in the role file, that module will not run.

Change this to more positive, like "Only modules included in the role file will be run."

Reviewer 2: I agree with Cameron - each role represents a binary itself. They're not really building blocks.

I think it would also be worth mentioning here that roles are the things you actually run, and perhaps link to the part in nubook where we say how to run a binary

Reviewer 3: > > If a module is not included in the role file, that module will not run.
> 
> Change this to more positive, like "Only modules included in the role file will be run."

Since it takes more than just the inclusion of a module in a role file for it to be run (e.g. appropriate messages also need to be emitted) we should say something like this

"Only modules included in the role file will have the opportunity to be run."

Author: I did hesitate on this line because of what @Bidski says. Although it might be good to explicitly mention that messages need to be emitted, rather than implying it.

Author: Both updated.

> When adding modules to the list in a role, use the syntax `namespace::module_name`. For example, if you want to include the [`SensorFilter` module](https://github.com/NUbots/NUbots/tree/master/module/input/SensorFilter), then check the namespace in [`SensorFilter.hpp`](https://github.com/NUbots/NUbots/blob/master/module/input/SensorFilter/src/SensorFilter.hpp). The namespace is `namespace module::input`, so we have `module::input::SensorFilter`. However, all modules have `module::` so we drop this part of the namespace. The line in the role is therefore `input::SensorFilter`.

Reviewer: This is assuming that at all times namespaces match the folder hierarchy.

Reviewer 2: To clarify Cameron's point - the folder hierarchy is actually what you need to specify. The actual namespace the module is in doesn't matter. This is perhaps too subtle a point to bother explaining on the overview page though - personally I would skate over the details

Reviewer 3: I would specify that its the folder hierarchy that needs to be specified. The first time someone has a module whose namespace differs from the folder hierarchy is going to be extremely confused because they are doing exactly what the docs said they should do.

Author: Ah, I always thought it was a namespace thing. One sure learns a lot writing documentation 😆

Author: Updated so it is correct.

The following is a table of roles that exist in the NUbots codebase and possible use cases for them.

> | Role Name            | What it contains                                                                                      | Usage                                                                                                                                                                                                                                                                   |

Reviewer 2: I don't think we should include what each role contains on here. That's subject to change which makes this too fragile in my opinion. I would say it's largely an implementation detail.

Author: Removed.

| -------------------- | ----------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `balllocalisation`   | Real robot sensors, Visual Mesh, ball detector and ball localisation                                  | Used to test and tune how well the robot detects balls on a field.                                                                                                                                                                                                      |

Reviewer 2: Did you consider removing the `Used` prefix from each of the usage comments? E.g. for this line making it `Testing and tuning how well the robot detects balls on a field.`

Author: Removed

| `cameratest`         | Real robot sensors and camera                                                                         | Used to see if the cameras are working and often used for [calibrating the cameras](/guides/main/camera-calibration).                                                                                                                                                   |
| `filter_playback`    | Filter sensor data, data playback and NUsight                                                         | Use for playing back raw sensor data and viewing the filtered data in NUsight.                                                                                                                                                                                          |

> | `firmwareinstaller`  | Firmware installer                                                                                    | Used to install firmware to the robot.                                                                                                                                                                                                                                  |

Reviewer 3: This still says nothing about what this module does. What firmware? (hint: it is used to install firmware on the CM740)

Author: "says nothing" - that is completely true because I had no idea what it did 😆 Should have included a review comment asking about it but by the time I finished, I forgot 😵 
You say 'used to', do we not need it anymore?

Reviewer 3: Read my comment again

Author: Ah, and with that I think it's time to sleep 😅

| `gamecontrollertest` | Real robot sensors, game controller and NUsight                                                       | Used for testing the connection between the robot and [GameController](https://github.com/RoboCup-Humanoid-TC/GameController).                                                                                                                                          |
| `gazebo`             | Gazebo robot sensors, motion modules and NUsight                                                      | Used to run motions with a robot in [Gazebo](http://gazebosim.org/).                                                                                                                                                                                                    |

> | `headbehtest`        | Real robot sensors and camera, vision modules, script modules, head controller and behaviour          | Used to test the behaviour of the head, particularly with balls and goals.                                                                                                                                                                                              |

Reviewer: ```suggestion
| `headbehtest`        | Real robot sensors and camera, vision modules, script modules, head controller and behaviour          | Used to test head behaviour, particularly with balls and goals.                                                                                                                                                                                              |
```

| `keyboardwalk`       | Real robot sensors, motions, keyboard walk and NUsight                                                | Used for demonstrating the robot manually with a keyboard, particularly in a setting without a field.                                                                                                                                                                   |

> | `keyboardwalkfake`   | As above but with a basic simulated robot (not webots)                                                | Used for testing the code without a real robot, potentially to make sure there are no runtime errors. Used to demonstrate basic functionality of the robot. Part of the [Getting Started](/guides/main/getting-started) example since it does not require a real robot. |

Reviewer 2: ```suggestion
| `keyboardwalkfake`   | As above but with a basic simulated robot (not webots)                                                | Used for testing the code for run-time issues without a real robot. Used to demonstrate basic functionality of the robot. Part of the [Getting Started](/guides/main/getting-started) example since it does not require a real robot. |
```

| `localisationtest`   | Localisation, localisation simulator and NUsight                                                      | Used to test localisation in a simulated environment.                                                                                                                                                                                                                   |

> | `lookatball`         | Real robot sensors, vision and motions                                                                | Used to test head behaviour where just the ball is the focus. By removing robot localisation, just the ball should be looked at.                                                                                                                                        |

Reviewer 2: ```suggestion
| `lookatball`         | Real robot sensors, vision and motions                                                                | Testing the robot's ability to track a ball with head behaviour.                                                                                                                                        |
```

> | `lookatgoal`         | Real robot sensors, localisation, vision and motions                                                  | Used to test head behaviour with localisation, which should make it look at the goals.                                                                                                                                                                                  |

Reviewer 2: ```suggestion
| `lookatgoal`         | Real robot sensors, localisation, vision and motions                                                  | Used to test head behaviour with localisation, which should make the robot look at the goals.                                                                                                                                                                                  |
```

> | `natnet`             | Real robot sensors, NUsight and NatNet                                                                | Used to get motion capture data from a [Motive](https://optitrack.com/software/motive/) motion capture system.                                                                                                                                                          |

Reviewer 2: glossary addition? :heart_eyes:

Author: Added, although can someone make sure my wording is ok for the NatNet entry? @Bidski?

> | `nusighttest`        | NUsight                                                                                               | Testing connection of NUsight, doesn't run anything that can send data.                                                                                                                                                                                                 |

Reviewer 2: ```suggestion
| `nusighttest`        | NUsight                                                                                               | Testing connection with NUsight. This role doesn't run anything that can send data.                                                                                                                                                                                                 |
```

| `webots_keyboard`    | `keyboardwalk` but with Webots instead of real robot sensors                                          | Used for easily testing motions in Webots.                                                                                                                                                                                                                              |
| `webots`             | Webots connection module                                                                              | Used for testing the connection between the robot code and Webots.                                                                                                                                                                                                      |
| `webotsrobocup`      | Identical to `robocup` but with webots instead of real robot sensors                                  | Used to play RoboCup games in the Webots simulator.                                                                                                                                                                                                                     |

> In most cases, `platform::darwin::HardwareIO` and `platform::Webots` can be used interchangably depending on if you are running on the real robot or the simulated robot in Webots. In some cases `platform::darwin::HardwareSimulator` can also be used, but this simulated robot is very basic and does not consider gravity - this may be useful if you want to see a motion run on the robot without the robot falling down.

Reviewer: Maybe mention that `HardwarIO` will crash if not run on real robot, whilst `Webots` and `HardwareSimulator` will be mostly fine.

Reviewer 3: You would also need to mention that `Webots` (and `Gazebo`) will only work if the appropriate simulator and controller/plugin are loaded correctly as well. `HardwareSimulator` should, in general, always run

Author: I put more explanation of the platforms but tbh this should be on a different page, like a platform page.

Author: I'm thinking leave it here for now and then move it over when a platform page is created? I'll add it to my list ✍️
